src/vivarium/framework/population/manager.py

```diff
@@ -399,19 +393,97 @@ def _create_simulants(
     # Context API #
     ###############
 
-    def get_population(self, untracked: bool) -> pd.DataFrame:
-        """Provides a copy of the full population state table.
+    def get_population_columns(self) -> list[str]:
+        """Get the list of columns in the population state table.
+
+        Returns
+        -------
+            The list of columns in the population state table.
+        """
+        return list(self._attribute_pipelines.keys())
+
+    def get_population(
+        self,
+        attributes: list[str] | Literal["all"],
+        untracked: bool,
+        index: pd.Index[int] | None = None,
+    ) -> pd.DataFrame:
+        """Provides a copy of the population state table.
 
         Parameters
         ----------
+        attributes
+            The attributes to include as the state table. If "all", all attributes are included.
         untracked
             Whether to include untracked simulants in the returned population.
+        index
+            The index of simulants to include in the returned population. If None,
+            all simulants are included (unless they are untracked and the untracked
+            argument is False).
 
         Returns
         -------
             A copy of the population table.
         """
-        pop = self._population.copy() if self._population is not None else pd.DataFrame()
-        if not untracked and "tracked" in pop.columns:
-            pop = pop[pop.tracked]
-        return pop
+
+        if self._population is None:
+            return pd.DataFrame()
+
+        idx = index if index is not None else self._population.index
+        if not untracked:
+            tracked_pipeline = self._attribute_pipelines["tracked"]
+            tracked = (
+                self._population.loc[idx, "tracked"]
+                if tracked_pipeline.is_simple
+                else tracked_pipeline(idx)
+            )
+            if not isinstance(tracked, pd.Series):
+                raise ValueError(
+                    "The 'tracked' attribute pipeline should return a pd.Series but instead "
+                    f"returned a {type(tracked)}."
+                )
+            idx = tracked[tracked].index
```

Comment:
This is correct but looks very weird. You could also do this, which might look clearer, and should be the same?
```python
idx = self._population[tracked].index
```